### PR TITLE
Taverns client info feature

### LINKED ISSUE #308: Create ClientInfo Class
## Description
Create a readonly struct called client info with the following data
- Client Guid
- Mod Version
- Mission Instance Name

The server must not allow connections with different versions of the mod for missions.

## Location
Changes shall exist in the Missions project
Branch base/target: [taverns-vertical-slice](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/taverns-vertical-slice)

## Related Issues
Subtask of #304 

## Requirements
- This class shall be serializable to a string by overriding the ToString() method.
- This class shall have a constructor that validates the given string
- This class shall have a constructor that constructs an object from the given string when valid

- The server shall only accept a connection that is using a single specified mod version.
- The [P2PClient](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/947dca56f116a9d09c3bae6634a3b2d1746d46c7/source/Missions/Network/LiteNetP2PClient.cs#L102) shall construct a ClientInfo object to send the following as a token
  - Client Guid
  - Mod Version
  - Instance name

- The [TestServer](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/947dca56f116a9d09c3bae6634a3b2d1746d46c7/source/TestServer/Server/MissionTestServer.cs#L52) shall deserialize the incoming ClientInfo class from the token field
- The [TestServer](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/947dca56f116a9d09c3bae6634a3b2d1746d46c7/source/TestServer/Server/MissionTestServer.cs#L52) shall use the received ClientInfo version data to validate the mod version
- The [TestServer](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/947dca56f116a9d09c3bae6634a3b2d1746d46c7/source/TestServer/Server/MissionTestServer.cs#L52) shall use the received ClientInfo version **Client Guid** and **Instance Name** to add and remove clients from instance on a connect or disconnect

## Definition of Done
- [ ] Class level documentation exists for every modified class
- [ ] Tests exist for every requirement above

### LINKED ISSUE #320: Convert IntroductionServer to .NET Standard
# For New Contributors
[How to contribute](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Getting-Started-as-a-Contributor)

## Description

Convert IntroductionServer to .NET Standard and move "Shared" into that class.
Create a .NET Core 6.0 for the console that runs the IntroductionServer as a library.

## Intended Design

N/A

## Location

The [tavern-vertical-slice branch](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/taverns-vertical-slice) after #319 is merged.

## Related Issues

N/A

## Requirements

N/A

## Additional information

N/A

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit tests exist for every method that does not require the game to be ran.

### PR DESCRIPTION
**Pull request by garrettluskey**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] All new classes have class-level documentation comments, if there are any at all
- [x] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:

## What is the current behaviour?
Any version of the mod could connect to the relay server allowing different mod versions to connect to eachother.

## What is the new behaviour?

Resolves #308
Resolves #320 

Added a class that sends the mod version from the client and the server validates if the client version matches the deployed version, accepting if it matches and rejecting if it doesn't.

## Other information

### CODE REVIEW

**Comment by Ceron257** [comment]
There are several generated PDB files in this merge request. They should be removed.

File: source/MissionTestMod/MissionTestMod.csproj [hunk 1]
```diff
@@ -13,6 +13,12 @@
     <FileAlignment>512</FileAlignment>
     <Deterministic>true</Deterministic>
   </PropertyGroup>
+  <PropertyGroup>
+    <ProduceReferenceAssembly>true</ProduceReferenceAssembly>
+    <Version>0.0.0</Version>
+    <AssemblyVersion>0.0.*</AssemblyVersion>
+    <Deterministic>false</Deterministic>
+</PropertyGroup>
```
Comment (Ceron257) [inline source/MissionTestMod/MissionTestMod.csproj]:
```suggestion
  </PropertyGroup>
```

File: source/SharedData/Extensions/StringBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        {
+            for (int i = 0; i < items.Length - 1; i++)
+            {
+                stringBuilder.Append(items[i].ToString());
+                stringBuilder.Append(delimiter);
+            }
+
+            stringBuilder.Append(items.Last());
+        }
+    }
```
Comment (Ceron257) [inline source/SharedData/Extensions/StringBuilderExtensions.cs]: This is a duplicate of `Common.Extensions.StringBuilderExtensions`.

**Review by EgardA — COMMENTED** [review]
Says "IntroducationServer" in some places, preferably have this as "IntroductionServer" as that is what the namespace seems to be called and prevents confusion later on.

File: source/IntroductionServer/Program.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
+﻿using IntroducationServer.Config;
+using IntroducationServer.Server;
```
Comment (EgardA) [inline source/IntroductionServer/Program.cs]:
```suggestion
using IntroductionServer.Server;
```

File: source/IntroductionServer/Program.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
+﻿using IntroducationServer.Config;
```
Comment (EgardA) [inline source/IntroductionServer/Program.cs]:
```suggestion
﻿using IntroductionServer.Config;
```

File: source/IntroductionServer/Program.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+using IntroducationServer.Server;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace IntroducationServer
```
Comment (EgardA) [inline source/IntroductionServer/Program.cs]:
```suggestion
namespace IntroductionServer
```

File: source/IntroductionServer/Server/MissionTestServer.cs [hunk 6]
```diff
@@ -1,39 +1,32 @@
...
-using System.Collections.Generic;
-using System.Diagnostics;
-using System.Linq;
-using System.Net.Sockets;
-using System.Net;
-using System.Text;
-using System.Threading.Tasks;
-using Missions.Network;
-using NLog;
+﻿using IntroducationServer.Config;
```
Comment (EgardA) [inline source/IntroductionServer/Server/MissionTestServer.cs]:
```suggestion
﻿using IntroductionServer.Config;
```

File: source/IntroductionServer/Server/MissionTestServer.cs [hunk 7]
```diff
@@ -1,39 +1,32 @@
...
 using LiteNetLib;
-using TestServer.Config;
+using NLog;
+using SharedData;
+using System;
+using System.Net;
+using System.Net.Sockets;
 
-namespace TestServer.Server
+namespace IntroducationServer.Server
```
Comment (EgardA) [inline source/IntroductionServer/Server/MissionTestServer.cs]:
```suggestion
namespace IntroductionServer.Server
```

File: source/IntroductionServer/Server/P2PPeer.cs [hunk 8]
```diff
@@ -6,7 +6,7 @@
 using System.Text;
 using System.Threading.Tasks;
 
-namespace Missions.Network
+namespace IntroducationServer.Server
```
Comment (EgardA) [inline source/IntroductionServer/Server/P2PPeer.cs]:
```suggestion
namespace IntroductionServer.Server
```

File: source/IntroductionServer/Server/PeerRegistry.cs [hunk 9]
```diff
@@ -8,16 +7,16 @@
 using System.Text;
 using System.Threading.Tasks;
 
-namespace TestServer.Server
+namespace IntroducationServer.Server
```
Comment (EgardA) [inline source/IntroductionServer/Server/PeerRegistry.cs]:
```suggestion
namespace IntroductionServer.Server
```

File: source/IntroductionServer/Config/NetworkConfiguration.cs [hunk 10]
```diff
@@ -1,7 +1,7 @@
 ﻿using System;
 using System.Net;
 
-namespace TestServer.Config
+namespace IntroducationServer.Config
```
Comment (EgardA) [inline source/IntroductionServer/Config/NetworkConfiguration.cs]:
```suggestion
namespace IntroductionServer.Config
```

File: source/SharedData/Extensions/StringBuilderExtensions.cs (hunk 2 above)
Comment (garrettluskey) [reply source/SharedData/Extensions/StringBuilderExtensions.cs]: Removed duplicate in Common

File: source/IntroductionServer/Program.cs (hunk 3 above)
Comment (garrettluskey) [reply source/IntroductionServer/Program.cs]: Name changed, comment no longer applicable

**Review by Ceron257 — APPROVED** [review]

**Pull request merged**